README.md
# `testpilot-ga`

[![testpilot-ga on npm](https://img.shields.io/npm/v/testpilot-ga.svg)](https://www.npmjs.com/package/testpilot-ga)

A library for event-driven data collection with Google Analytics for [Firefox Test Pilot](https://testpilot.firefox.com) experiments. Aims to follow [Test Pilot standards](https://github.com/mozilla/testpilot/blob/master/docs/experiments/ga.md) for reporting, automatically respects Do Not Track settings.

## Table of contents

- [API Documentation](docs/api.md)
- [Example WebExtension](example)
- [Changelog](docs/changelog.md)
- [License](LICENSE)
- [Code of conduct](docs/CODE_OF_CONDUCT.md)

## Usage

1. Add the [`testpilot-ga` package](https://www.npmjs.com/package/testpilot-ga)
   using your package manager:

   ```
   npm install --save testpilot-ga
   ```

   or

   ```
   yarn add testpilot-ga
   ```

2. Import and create a singleton instance of the `TestPilotGA` class exported by
   `testpilot-ga`:

   ```js
   import TestPilotGA from 'testpilot-ga';

   const analytics = new TestPilotGA({
     aid: 'test-experiment@mozilla.com',
     an: 'Test Experiment',
     av: '1.0.0',
     cd19: 'production',
     ds: 'addon',
     tid: 'UA-71632928-4'
   });
   ```

3. Use that instance to send events to Google Analytics:

   ```js
   analytics
     .sendEvent('ec_value', 'ea_value', {
       cd1: 'cd1_value',
       cm1: 'cm1_value'
     })
     .then(response => {
       console.log('Event successfully sent', response)
     })
     .catch((response, err) => {
       console.error('Event failed while sending', response, err)
     });
   ```

   Which produces this request:

   ```
   POST /collect HTTP/1.1
   Host: www.google-analytics.com

   v=1&t=event&tid=UA-71632928-4&cid=00a9e969-362b-4cee-b707-c63e17a4e71a&an=Test%20Experi
   ment&aid=test-experiment%40mozilla.com&aiid=testpilot&aip=1&av=1.0.0&ds=addon&ua=Mozill
   a%2F5.0%20(Macintosh%3B%20Intel%20Mac%20OS%20X%2010_12_5)%20AppleWebKit%2F537.36%20(KHT
   ML%2C%20like%20Gecko)%20Chrome%2F59.0.3071.115%20Safari%2F537.36&ul=en-US&z=15004209550
   09&ec=catname&ea=eventname&cd1=cd1val1&cm1=cm1val1&cd19=production&cd20=release
   ```

## More information

- IRC: #testpilot on irc.mozilla.org


docs/changelog.md
# Changelog

## `0.3.0`

Released 2017/07/20

- Includes `cd19` (experiment release channel) in all events.
- Includes `cd20` (Firefox release channel) in all events from WebExtensions.
- Logs the full contents of each event before serializing.


src/index.js
const LOCALSTORAGE_CID = "testpilot_ga__cid";

export default class TestPilotGA {
  static defaultOptions = {
    aid: null,
    aiid: "testpilot",
    aip: "1",
    av: null,
    ds: "addon",
    t: "event",
    uid: null,
    v: "1",
    xid: null,
    xvar: null
  };

  static requiredOptions = ["an", "ds", "tid"];

  constructor(options) {
    this.debug = options.debug || false;
    this.setOptions(options);
    this.validateOptions();
  }

  setOptions(options) {
    const x = 1;
    const allOptions = Object.assign({}, TestPilotGA.defaultOptions, options);
    Object.entries(allOptions).forEach(([key, value]) => (this[key] = value));
  }

  validateOptions(options) {
    const missingOptions = TestPilotGA.requiredOptions.reduce((accum, opt) => {
      if (!this.hasOwnProperty(opt)) {
        accum.push(opt);
      }
      return accum;
    }, []);
    if (missingOptions.length) {
      console.error(`Missing required options: ${missingOptions.join(", ")}`);
    }
  }

  makeUUID() {
    // From https://stackoverflow.com/a/2117523
    return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, c =>
      (c ^
        (crypto.getRandomValues(new Uint8Array(1))[0] &
          (15 >> (c / 4)))).toString(16)
    );
  }

  getCID() {
    let cid = window.localStorage.getItem(LOCALSTORAGE_CID);
    if (!cid) {
      cid = this.makeUUID();
      window.localStorage.setItem(LOCALSTORAGE_CID, cid);
    }
    return cid;
  }

  getParams(eventParams) {
    const { an, aid, aiid, aip, av, ds, t, tid, uid, v, xid, xvar } = this;
    const params = Object.assign(
      { an, aid, aiid, aip, av, ds, t, tid, uid, v, xid, xvar },
      {
        cid: this.getCID(),
        ua: navigator.userAgent,
        ul: navigator.language,
        z: Date.now()
      },
      eventParams
    );
    Object.keys(params).forEach(paramName => {
      if (params[paramName] === null) {
        delete params[paramName];
      }
    });
    return params;
  }

  serializeObject(obj) {
    return Object.keys(obj)
      .reduce((accum, param) => {
        accum.push(
          `${encodeURIComponent(param)}=${encodeURIComponent(obj[param])}`
        );
        return accum;
      }, [])
      .join("&");
  }

  requestBody(eventParams) {
    return this.serializeObject(this.getParams(eventParams));
  }

  requestURI() {
    const hostname = "www.google-analytics.com";
    return `https://${hostname}/${this.debug ? "debug/" : ""}collect`;
  }

  sendEvent(ec, ea, params = {}) {
    const eventParams = Object.assign({ ec, ea }, params);
    const requestBody = this.requestBody(eventParams);
    const requestUri = this.requestURI();
    return new Promise((resolve, reject) => {
      if (navigator.doNotTrack === "1") {
        reject(null, Error("Metrics aborted due to DNT."));
      } else {
        const req = new window.XMLHttpRequest();
        req.open("POST", requestUri);
        req.onload = function() {
          if (req.status < 400) {
            resolve(req);
          } else {
            reject(req, Error(req.statusText));
          }
        };
        req.onerror = function() {
          reject(req, Error("Network Error"));
        };
        req.send(requestBody);
      }
    });
  }
}


src/TestPilotGA.js
const CHANNELS = {
  FIREFOX_NIGHTLY: "nightly",
  FIREFOX_AURORA: "release",
  FIREFOX_ESR: "esr",
  FIREFOX_ESR_NEXT: "esr",
  LATEST_FIREFOX_DEVEL_VERSION: "developer",
  LATEST_FIREFOX_RELEASED_DEVEL_VERSION: "developer",
  LATEST_FIREFOX_VERSION: "release"
};
const LOCALSTORAGE_CID = "testpilot_ga__cid";
const PRODUCT_DETAILS_URL =
  "https://product-details.mozilla.org/1.0/firefox_versions.json";

export default class TestPilotGA {
  static defaultOptions = {
    aid: null,
    aiid: "testpilot",
    aip: "1",
    av: null,
    cd19: "dev",
    cd20: null,
    ds: "addon",
    t: "event",
    uid: null,
    v: "1",
    xid: null,
    xvar: null
  };

  static requiredOptions = ["an", "ds", "tid"];

  constructor(options) {
    this.debug = options.debug || false;
    this.setOptions(options);
    this.getChannel();
    this.validateOptions();
  }

  getChannel() {
    if (typeof browser !== "undefined") {
      Promise.all([
        this.getProductDetails(),
        browser.runtime.getBrowserInfo()
      ]).then(([productDetails, browserInfo]) => {
        const versionMap = this.getVersionMap(productDetails);
        this.cd20 = this.getChannelConstant(browserInfo, versionMap);
      });
    }
  }

  getVersionMap(productDetails) {
    const versionMap = {};
    Object.entries(productDetails).forEach(([key, value]) => {
      if (value && value.length) {
        versionMap[value] = key;
      }
      const abbreviated = value.match(/^[^a-z]+/);
      if (abbreviated && abbreviated.length) {
        versionMap[abbreviated[0]] = key;
      }
    });
    return versionMap;
  }

  getChannelConstant(browserInfo, versionMap) {
    return browserInfo &&
    browserInfo.version &&
    versionMap.hasOwnProperty(browserInfo.version) &&
    CHANNELS.hasOwnProperty(versionMap[browserInfo.version])
      ? CHANNELS[versionMap[browserInfo.version]]
      : "other";
  }

  getProductDetails() {
    return new Promise((resolve, reject) => {
      const req = new window.XMLHttpRequest();
      req.open("GET", PRODUCT_DETAILS_URL);
      req.onload = function() {
        if (req.status < 400) {
          resolve(JSON.parse(req.response));
        } else {
          reject(req, Error(req.statusText));
        }
      };
      req.onerror = function() {
        reject(req, Error("Network Error"));
      };
      req.send();
    });
  }

  setOptions(options) {
    const x = 1;
    const allOptions = Object.assign({}, TestPilotGA.defaultOptions, options);
    Object.entries(allOptions).forEach(([key, value]) => (this[key] = value));
  }

  validateOptions(options) {
    const missingOptions = TestPilotGA.requiredOptions.reduce((accum, opt) => {
      if (!this.hasOwnProperty(opt)) {
        accum.push(opt);
      }
      return accum;
    }, []);
    if (missingOptions.length) {
      console.error(`Missing required options: ${missingOptions.join(", ")}`);
    }
  }

  makeUUID() {
    // From https://stackoverflow.com/a/2117523
    return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, c =>
      (c ^
        (crypto.getRandomValues(new Uint8Array(1))[0] &
          (15 >> (c / 4)))).toString(16)
    );
  }

  getCID() {
    let cid = window.localStorage.getItem(LOCALSTORAGE_CID);
    if (!cid) {
      cid = this.makeUUID();
      window.localStorage.setItem(LOCALSTORAGE_CID, cid);
    }
    return cid;
  }

  getParams(eventParams) {
    const {
      an,
      aid,
      aiid,
      aip,
      av,
      cd19,
      cd20,
      ds,
      t,
      tid,
      uid,
      v,
      xid,
      xvar
    } = this;
    const params = Object.assign(
      { an, aid, aiid, aip, av, cd19, cd20, ds, t, tid, uid, v, xid, xvar },
      {
        cid: this.getCID(),
        ua: navigator.userAgent,
        ul: navigator.language,
        z: Date.now()
      },
      eventParams
    );
    Object.keys(params).forEach(paramName => {
      if (params[paramName] === null) {
        delete params[paramName];
      }
    });
    return params;
  }

  serializeObject(obj) {
    return Object.keys(obj)
      .reduce((accum, param) => {
        accum.push(
          `${encodeURIComponent(param)}=${encodeURIComponent(obj[param])}`
        );
        return accum;
      }, [])
      .join("&");
  }

  requestBody(eventParams) {
    const allParams = this.getParams(eventParams);
    return {
      allParams,
      requestBody: this.serializeObject(allParams)
    };
  }

  requestURI() {
    const hostname = "www.google-analytics.com";
    return `https://${hostname}/${this.debug ? "debug/" : ""}collect`;
  }

  sendEvent(ec, ea, params = {}) {
    const eventParams = Object.assign({ ec, ea }, params);
    const { allParams, requestBody } = this.requestBody(eventParams);
    console.log(`Sending '${ec}' '${ea}':`, allParams);
    const requestUri = this.requestURI();
    return new Promise((resolve, reject) => {
      if (navigator.doNotTrack === "1") {
        reject("Metrics not sent due to DNT.");
      } else {
        const req = new window.XMLHttpRequest();
        req.open("POST", requestUri);
        req.onload = function() {
          if (req.status < 400) {
            resolve(req);
          } else {
            reject(`Request error: ${req.statusText}`);
          }
        };
        req.onerror = function() {
          reject(`Request error: ${req.status}`);
        };
        req.send(requestBody);
      }
    });
  }
}
